src/golem-network/golem-network.ts

```diff
@@ -314,20 +315,35 @@ export class GolemNetwork {
     }
   }
 
+  private async startDisconnect() {
+    try {
+      await Promise.allSettled(this.cleanupTasks.map((task) => task()));
+      await this.storageProvider.close();
+      await this.yagna.disconnect();
+
+      this.services.proposalCache.flushAll();
+    } catch (err) {
+      this.logger.error("Error while disconnecting", err);
+      throw err;
+    } finally {
+      this.events.emit("disconnected");
+      this.hasConnection = false;
+    }
+  }
+
   /**
    * "Disconnects" from the Golem Network
    *
    * @return Resolves when all shutdown steps are completed
    */
   async disconnect() {
-    await Promise.allSettled(this.cleanupTasks.map((task) => task()));
-    await this.storageProvider.close();
-    await this.yagna.disconnect();
-
-    this.services.proposalCache.flushAll();
-
-    this.events.emit("disconnected");
-    this.hasConnection = false;
+    if (this.disconnectPromise) {
+      return this.disconnectPromise;
+    }
+    this.disconnectPromise = this.startDisconnect().finally(() => {
+      this.disconnectPromise = undefined;
```

Comment:
why is this needed? this means that running disconnect sequentially will not be idempotent
```ts
await glm.disconnect();
await glm.disconnect(); // runs disconnect again causes another attempt to releas allocation etc.
```
so maybe it should be `undefined` only when the user run the `connect` again?

Reply:
`cleanupTasks` should be set to `null` after disconnect finishes, this is a bug